QR Code 발급 및 인증 기능 구현 완료

cxxxtxxyxx opened the pull request:
- #22 

### 추가 구현 필요
- [x] JWT 연동 및 검증
- [x] UnitTest 작성
    - [x] ControllerTest
    - [x] ServiceTest

cxxxtxxyxx committed:
Feat: 필요한 Repository 생성 (#22)

Invitation, Visitor Repository 생성

cxxxtxxyxx committed:
Chore: QR코드 라이브러리 의존성 추가 (#22)

zxing 의존성 추가

cxxxtxxyxx committed:
Feat: QR코드 매니저 클래스 구현 (#22)

QrCodeManager 클래스 생성 및 구현
QrPayload 클래스 생성

cxxxtxxyxx committed:
Feat: InvitationId를 가져오는 메서드 구현 (#22)

VisitorId를 통해 해당 방문자의 초대장 Id를 조회

cxxxtxxyxx committed:
Feat: 초대장 시작 및 종료 시간을 가져오는 메서드 구현 (#22)

QR 생성 및 유효성 검사를 위해 초대장 시작 및 종료 시간을 조회

cxxxtxxyxx committed:
Feat: QR생성 및 validate 위임 메서드 구현 (#22)

VisitationService에서 QrCodeManager를 사용하여 QR 관리

cxxxtxxyxx committed:
Feat: VisitationFacadeService 구현 (#22)

Service, Repository의 의존성을 1대1로 관리하기 위해 Facade 객체 구현

cxxxtxxyxx committed:
Feat: VisitationErrorCode 구현 (#22)

HttpStatus 및 Message 관리 객체

cxxxtxxyxx committed:
Feat: VisitationController 구현 (#22)

cxxxtxxyxx committed:
Modify: InvitationRepository 메서드 제거 (#22)

사용하지 않는 불필요한 메서드 제거

cxxxtxxyxx committed:
Typo: 메서드 명 수정 (#22)

컨벤션에 맞도록 메서드 명 수정
get -> create

cxxxtxxyxx committed:
Typo: 줄바꿈 추가 (#22)

컨벤션에 맞도록 줄바꿈 추가

cxxxtxxyxx committed:
Test: QRCode 생성 요청 테스트 구현 (#22)

VisitationController createQrCode에 대한 MVC 테스트 구현

hyunsb commented on src/main/java/com/livable/server/visitation/service/InvitationService.java:
```diff
@@ -0,0 +1,29 @@
+                .startDate(invitation.getStartDate())
+                .endDate(invitation.getEndDate())
+                .build();
+    }
```
요거 `InvitationTimeDTO` 안에 정적 팩토리 메서드로 처리하면 더 깔끔할 거 같아요
```java
public static InvitationTimeDTO from(Invitation invitation) {
    return InvitationTimeDTO.builder()
    ...
    ;
}
```

hyunsb commented on src/main/java/com/livable/server/visitation/domain/VisitationErrorCode.java:
```diff
@@ -0,0 +1,22 @@
+
+    private final HttpStatus httpStatus;
+    private final String message;
+}
```
이거 설계가 참 좋은 거 같습니다 👍

cxxxtxxyxx committed:
Test: QRCode 생성 요청 실패 테스트 구현 (#22)

VisitationController createQrCode에 대한 MVC 테스트 구현

cxxxtxxyxx committed:
Test: QRCode 생성 요청 실패(400) 테스트 구현 (#22)

cxxxtxxyxx committed:
Test: QRCode 생성 요청 실패(500) 테스트 구현 (#22)

cxxxtxxyxx committed:
Refactor: 테스트 메서드 네이밍 컨트롤러와 통일 (#22)

cxxxtxxyxx committed:
Test: 테스트용 모킹 객체 생성 (#22)

RequestBodyDto 객체에 @Builder 어노테이션을 붙이는 대신 상속 모킹 객체를 생성해서 대신 사용

cxxxtxxyxx committed:
Test: QRCode 인증 성공 테스트 구현 (#22)

cxxxtxxyxx committed:
Test: QRCode 인증 실패(400) 테스트 구현 (#22)

cxxxtxxyxx committed:
Test: findInvitationTime 성공 테스트 구현 (#22)

cxxxtxxyxx committed:
Test: findInvitationTime 실패 테스트 구현 (#22)

cxxxtxxyxx committed:
Test: createQrCode 성공 테스트 구현 (#22)

cxxxtxxyxx committed:
Test: validateQrCode 성공 테스트 구현 (#22)

VisitationService 테스트

cxxxtxxyxx committed:
Test: VisitorService.findInvitationId 성공 테스트 구현 (#22)

cxxxtxxyxx committed:
Style: 불필요한 Import문 제거 (#22)

InvitationServiceTest 내부에 MockBean import문 제거

cxxxtxxyxx committed:
Test: VisitorService.findInvitationId 실패 테스트 구현 (#22)

cxxxtxxyxx committed:
Refactor: Qr생성 Validation 추가 및 수정 (#22)

cxxxtxxyxx committed:
Fix: LocalDateTime 역직렬화를 위한 코드 추가 (#22)

objectMapper.registerModule(new JavaTimeModule()).writeValueAsString(expirationPeriodMap)로 수정

cxxxtxxyxx committed:
Test: QrCodeManager.createQrCode 성공 테스트 구현 (#22)

cxxxtxxyxx committed:
Test: QrCodeManager.createQrCode 실패 테스트_1 구현 (#22)

cxxxtxxyxx committed:
Test: QrCodeManager.createQrCode 실패 테스트_2 구현 (#22)

cxxxtxxyxx commented on src/main/java/com/livable/server/visitation/controller/VisitationController.java:
```diff
@@ -0,0 +1,36 @@
+    @GetMapping("/qr")
+    public ResponseEntity<ApiResponse.Success<Object>> createQrCode() {
+
+        Long visitorId = 1L;
+        VisitationResponse.InvitationTimeDto invitationTime = visitationFacadeService.findInvitationTime(visitorId);
+        String base64QrCode = visitationFacadeService.createQrCode(invitationTime.getStartDateTime().minusHours(1), invitationTime.getEndDateTime().plusHours(1));
+
+        return ApiResponse.success(base64QrCode, HttpStatus.OK);
```
이 부분은 JWT 연동 이후 토큰에서 값 가져오도록 수정할 것 같습니다!

cxxxtxxyxx committed:
Refactor: Controller 메서드에서는 Facade 객체로 위임만 하도록 리팩터링 (#22)

Facade객체에서 다양한 Service에 대한 의존을 통해 로직 처리

cxxxtxxyxx committed:
Refactor: 로직 수정에 따른 테스트코드 리팩터링 (#22)

cxxxtxxyxx committed:
Modify: 불필요한 예외 테스트 제거 (#22)

cxxxtxxyxx committed:
Test: VisitationFacadeService.createQrCode 성공 테스트 구현 (#22)

cxxxtxxyxx committed:
Test: VisitationFacadeService.validateQrCode 성공 테스트 구현 (#22)

cxxxtxxyxx committed:
Feat: 쿼리 줄이기 위해 JPQL 직접 작성 (#22)

cxxxtxxyxx committed:
Fix: Repository 테스트 실패 제거 (#22)

hyunsb commented on src/main/java/com/livable/server/visitation/domain/QrCodeManager.java:
```diff
@@ -0,0 +1,155 @@
+
+            Map<EncodeHintType, Object> encodeHints = getEncodeHints();
+
+            QRCodeWriter qrCodeWriter = new QRCodeWriter();
```
해당 메서드는 호출될 때마다 QRCodeWriter 객체가 생성될 거 같아요
메서드 내에서 특정 객체를 new 해버리면 테스트하기 어려울 거 같습니다!
